### ISSUE TITLE
Getting nan after keeping clip grad in torch

### ISSUE THREAD

**Issue by shiva1393** [mravanelli/pytorch-kaldi#154]
Hi Sir,
While training LSTM for 1000 hours(Librispeech) ,Iam able to train upto 1 epoch later getting nan. I tried with use torch.nn.utils.clip_grad_norm_(nns[net].parameters(), 0.1) but i got nan in loss.
ep=00 tr=['train_960'] loss=1.554 ....
ep=01 tr=['train_960'] loss=nan ....
This same thing happened for Li-GRU network.
Any suggestions? to overcome this issue.

**Comment by mravanelli** [comment]
Hi,
could you post the LSTM config file?

**Comment by shiva1393** [comment]
[cfg_proto]
cfg_proto=proto/global.proto
cfg_proto_chunk=proto/global_chunk.proto

[exp]
cmd=
run_nn_script=run_nn
out_folder=exp_librispeech/libri_LSTM_fmllr_1000hour
seed=1234
use_cuda=True
multi_gpu=False
save_gpumem=False
N_epochs_tr=24

[dataset1]
data_name=train_960
fea:fea_name=fbank
    fea_lst=/home/fbank-1000/train_960/feats.scp
    fea_opts=apply-cmvn --utt2spk=ark:/home/train_960/utt2spk  ark:/home/cmvn_train_960.ark ark:- ark:- | add-deltas --delta-order=0 ark:- ark:- |
    cw_left=0
    cw_right=0

    
lab:lab_name=lab_cd
    lab_folder=/home/tri6b_ali_960/
    lab_opts=ali-to-pdf 
    lab_count_file=auto
    lab_data_folder=/home/train_960/
    lab_graph=/home/tri6b/graph_tgsmall/

N_chunks=20
        
[dataset2]
data_name=dev_clean
fea:fea_name=fbank
    fea_lst=/home/fbank-1000/dev_clean/feats.scp
    fea_opts=apply-cmvn --utt2spk=ark:/home/dev_clean/utt2spk  ark:/home/cmvn_dev_clean.ark ark:- ark:- | add-deltas --delta-order=0 ark:- ark:- |
    cw_left=0
    cw_right=0


lab:lab_name=lab_cd
    lab_folder=/home/tri6b_ali_dev_clean_100
    lab_opts=ali-to-pdf 
    lab_count_file=auto
    lab_data_folder=/home/dev_clean/
    lab_graph=/home/tri6b/graph_tgsmall/

N_chunks=10

[dataset3]
data_name=test_clean
fea:fea_name=fbank
    fea_lst=/home/fbank-1000/test_clean/feats.scp
    fea_opts=apply-cmvn --utt2spk=ark:/home/test_clean/utt2spk  ark:/home/cmvn_test_clean.ark ark:- ark:- | add-deltas --delta-order=0 ark:- ark:- |
    cw_left=0
    cw_right=0


lab:lab_name=lab_cd
    lab_folder=/home/tri6b_ali_test_clean_100/
    lab_opts=ali-to-pdf 
    lab_count_file=auto
    lab_data_folder=/home/test_clean/
    lab_graph=/home/tri6b/graph_tgsmall/

N_chunks=8

        
[data_use]
train_with=train_960
valid_with=dev_clean
forward_with=test_clean


[batches]
batch_size_train=16
max_seq_length_train=500
increase_seq_length_train=True
start_seq_len_train=100
multply_factor_seq_len_train=2
batch_size_valid=8
max_seq_length_valid=1000

[architecture1]
arch_name = LSTM_layers_first
arch_proto = proto/LSTM.proto
arch_library = neural_networks_ref4
arch_class = LSTM
arch_pretrain_file = none
arch_freeze = False
arch_seq_model = True
lstm_lay = 550,550
lstm_drop = 0.2,0.2
lstm_use_laynorm_inp = False
lstm_use_batchnorm_inp = False
lstm_use_laynorm = False,False
lstm_use_batchnorm = True,True
lstm_bidir = True
lstm_act = tanh,tanh
lstm_orthinit=True

arch_lr = 0.0016
arch_halving_factor = 0.5
arch_improvement_threshold = 0.001
arch_opt = rmsprop
opt_momentum = 0.0
opt_alpha = 0.95
opt_eps = 1e-8
opt_centered = False
opt_weight_decay = 0.0

[architecture2]
arch_name = LSTM_layers_second
arch_proto = proto/LSTM.proto
arch_library = neural_networks_ref4
arch_class = LSTM
arch_pretrain_file = none
arch_freeze = False
arch_seq_model = True
lstm_lay = 550
lstm_drop = 0.2
lstm_use_laynorm_inp = False
lstm_use_batchnorm_inp = False
lstm_use_laynorm = False
lstm_use_batchnorm = True
lstm_bidir = True
lstm_act = tanh
lstm_orthinit=True

arch_lr = 0.0016
arch_halving_factor = 0.5
arch_improvement_threshold = 0.001
arch_opt = rmsprop
opt_momentum = 0.0
opt_alpha = 0.95
opt_eps = 1e-8
opt_centered = False
opt_weight_decay = 0.0

[architecture3]
arch_name = LSTM_layers_third
arch_proto = proto/LSTM.proto
arch_library = neural_networks_ref4
arch_class = LSTM
arch_pretrain_file = none
arch_freeze = False
arch_seq_model = True
lstm_lay = 550
lstm_drop = 0.2
lstm_use_laynorm_inp = False
lstm_use_batchnorm_inp = False
lstm_use_laynorm = False
lstm_use_batchnorm = True
lstm_bidir = True
lstm_act = tanh
lstm_orthinit=True

arch_lr = 0.0016
arch_halving_factor = 0.5
arch_improvement_threshold = 0.001
arch_opt = rmsprop
opt_momentum = 0.0
opt_alpha = 0.95
opt_eps = 1e-8
opt_centered = False
opt_weight_decay = 0.0

[architecture4]
arch_name = LSTM_layers_fourth
arch_proto = proto/LSTM.proto
arch_library = neural_networks_ref4
arch_class = LSTM
arch_pretrain_file = none
arch_freeze = False
arch_seq_model = True
lstm_lay = 550
lstm_drop = 0.2
lstm_use_laynorm_inp = False
lstm_use_batchnorm_inp = False
lstm_use_laynorm = False
lstm_use_batchnorm = True
lstm_bidir = True
lstm_act = tanh
lstm_orthinit=True

arch_lr = 0.0016
arch_halving_factor = 0.5
arch_improvement_threshold = 0.001
arch_opt = rmsprop
opt_momentum = 0.0
opt_alpha = 0.95
opt_eps = 1e-8
opt_centered = False
opt_weight_decay = 0.0


[architecture5]
arch_name = MLP_layers_last
arch_proto = proto/MLP.proto
arch_library = neural_networks_ref4
arch_class = MLP
arch_pretrain_file = none
arch_freeze = False
arch_seq_model = False
dnn_lay = N_out_lab_cd
dnn_drop = 0.0
dnn_use_laynorm_inp = False
dnn_use_batchnorm_inp = False
dnn_use_batchnorm = False
dnn_use_laynorm = False
dnn_act = softmax
arch_lr = 0.0004
arch_halving_factor = 0.5
arch_improvement_threshold = 0.001
arch_opt = rmsprop
opt_momentum = 0.0
opt_alpha = 0.95
opt_eps = 1e-8
opt_centered = False
opt_weight_decay = 0.0


[model]
model_proto = proto/model.proto
model = out_dnn1=compute(LSTM_layers_first,fbank)
	out_dnn2=compute(LSTM_layers_second,out_dnn1)
	out_dnn3=compute(LSTM_layers_third,out_dnn2)
	out_dnn4=compute(LSTM_layers_fourth,out_dnn3)
	out_dnn5=compute(MLP_layers_last,out_dnn4)
	loss_final=cost_nll(out_dnn5,lab_cd)
	err_final=cost_err(out_dnn5,lab_cd)

[forward]
forward_out = out_dnn5
normalize_posteriors = True
normalize_with_counts_from = lab_cd
save_out_file = False
require_decoding = True

[decoding]
decoding_script_folder=kaldi_decoding_scripts/
decoding_script=decode_dnn.sh
decoding_proto=proto/decoding.proto
min_active=200
max_active=7000
max_mem=50000
beam=20.0
latbeam=12.0
acwt=0.10
max_arcs=-1
skip_scoring=false
scoring_script=/home/kaldi/egs/librispeech/s5/local/score.sh
scoring_opts="--min-lmwt 4 --max-lmwt 23"
norm_vars=False

**Comment by mravanelli** [comment]
Hi,
why you are defining a different architecture for each layer? This
shouldn't be the problem, but it is more clean to define it as reported
here:

https://github.com/mravanelli/pytorch-kaldi/blob/master/cfg/Librispeech_baselines/libri_LSTM_fmllr.cfg

The main difference is that in the given example we only do training with
the 100h clean part. Could you try to see if with the 100h subset
everything is working. I don't see nothing weird in the config file...

Mirco

**Comment by shiva1393** [comment]
For Li-GRU and LSTM both i tried for 100 hours it worked fine.For Li-GRU iam able to produce same results which mentioned in paper.But when i tried Li-GRU and LSTM for 1000 hours iam getting loss=nan at second or third epoch.I tried same cfg file (libri_LSTM_fmllr.cfg) which you mentioned but iam getting nan. For Transfer Learning purpose iam using network like this.

**Comment by mravanelli** [comment]
I see. This means that inside the librispeech 1000h there are some bad data
that messes up the acoustic model. One idea is to try to print the
sentence_ids of the batches that causes the problem. This way we can figure
out what is the problematic sentence. Also, have you tried to change the
seed and see if it works fine (with a different seed we shuffle the data in
a different way and maybe we avoid this unfortunate condition)?

Mirco

**Comment by TParcollet** [comment]
I would also consider reducing slightly the learning rate to see if this is not the problem first. Indeed, between 100H and 900H, the model is trained with a lot more data at the same learning rate/sequence size.

**Comment by shiva1393** [comment]
Thank you sir , I will try this three ways changing seed, sentence_ids and learning rate. If i able to train i will post which method works for 1000h.

**Comment by mravanelli** [comment]
Please, keep us updated!

**Comment by wpy3345** [comment]
Has the problem been solved?
